File: rail-module/rail-service/src/main/java/com/hillayes/rail/service/CategoryService.java
package com.hillayes.rail.service;

import com.hillayes.commons.Strings;
import com.hillayes.commons.jpa.Page;
import com.hillayes.exception.common.MissingParameterException;
import com.hillayes.exception.common.NotFoundException;
import com.hillayes.rail.domain.Category;
import com.hillayes.rail.domain.CategoryGroup;
import com.hillayes.rail.domain.CategoryStatistics;
import com.hillayes.rail.domain.CategorySelector;
import com.hillayes.rail.errors.CategoryAlreadyExistsException;
import com.hillayes.rail.errors.CategoryGroupAlreadyExistsException;
import com.hillayes.rail.repository.AccountRepository;
import com.hillayes.rail.repository.CategoryGroupRepository;
import com.hillayes.rail.repository.CategoryRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.time.Instant;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

@ApplicationScoped
@Transactional
@RequiredArgsConstructor
@Slf4j
public class CategoryService {
    private final CategoryGroupRepository categoryGroupRepository;
    private final CategoryRepository categoryRepository;
    private final AccountRepository accountRepository;

    /**
     * Returns the selected page of category groups for the specified user; in name order.
     *
     * @param userId The user ID.
     * @param pageIndex The zero-based, page number.
     * @param pageSize The max number of groups per page.
     * @return The selected page of category groups.
     */
    public Page<CategoryGroup> getCategoryGroups(UUID userId, int pageIndex, int pageSize) {
        log.info("Listing category groups [userId: {}, page: {}, pageSize: {}]", userId, pageIndex, pageSize);
        Page<CategoryGroup> result = categoryGroupRepository.findByUserId(userId, pageIndex, pageSize);

        if (log.isDebugEnabled()) {
            log.debug("Listing category groups [userId: {}, page: {}, pageSize: {}, size: {}, totalCount: {}]",
                userId, pageIndex, pageSize, result.getContentSize(), result.getTotalCount());
        }
        return result;
    }

    public CategoryGroup getCategoryGroup(UUID userId, UUID groupId) {
        log.info("Getting category group [userId: {}, groupId: {}]", userId, groupId);
        return categoryGroupRepository.findByIdOptional(groupId)
            .filter(category -> category.getUserId().equals(userId))
            .orElseThrow(() -> new NotFoundException("CategoryGroup", groupId));
    }

    public CategoryGroup createCategoryGroup(UUID userId, String name, String description) {
        log.info("Creating category group [userId: {}, name: {}]", userId, name);

        String newName = Strings.trimOrNull(name);
        if (newName == null) {
            throw new MissingParameterException("CategoryGroup.name");
        }

        categoryGroupRepository.findByUserAndName(userId, newName)
            .ifPresent(existing -> { throw new CategoryGroupAlreadyExistsException(existing); });

        CategoryGroup group = CategoryGroup.builder()
            .userId(userId)
            .name(newName)
            .description(Strings.trimOrNull(description))
            .build();
        return categoryGroupRepository.save(group);
    }

    public CategoryGroup updateCategoryGroup(UUID userId, UUID groupId,
                                             String name, String description) {
        log.info("Updating category group [userId: {}, groupId: {}, name: {}]", userId, groupId, name);
        String newName = Strings.trimOrNull(name);
        if (newName == null) {
            throw new MissingParameterException("CategoryGroup.name");
        }

        CategoryGroup group = getCategoryGroup(userId, groupId);

        categoryGroupRepository.findByUserAndName(userId, newName)
            .filter(existing -> !existing.getId().equals(groupId))
            .ifPresent(existing -> { throw new CategoryGroupAlreadyExistsException(existing); });

        group.setName(newName);
        group.setDescription(Strings.trimOrNull(description));
        return categoryGroupRepository.save(group);
    }

    public CategoryGroup deleteCategoryGroup(UUID userId, UUID groupId) {
        log.info("Deleting category group [userId: {}, groupId: {}]", userId, groupId);
        CategoryGroup group = getCategoryGroup(userId, groupId);

        categoryGroupRepository.delete(group);
        return group;
    }

    /**
     * Deletes all category groups (and their categories and selectors) for the specified user.
     *
     * @param userId The user ID.
     */
    public void deleteAllCategoryGroups(UUID userId) {
        log.info("Deleting all category groups for user [userId: {}]", userId);
        categoryGroupRepository.deleteByUserId(userId);
    }

    /**
     * Returns the selected page of categories for the identified category group; in name order.
     *
     * @param groupId The category group ID.
     * @param pageIndex The zero-based, page number.
     * @param pageSize The max number of groups per page.
     * @return The selected page of category groups.
     */
    public Page<Category> getCategories(UUID userId, UUID groupId, int pageIndex, int pageSize) {
        log.info("Listing categories for group [userId: {}, groupId: {}, page: {}, pageSize: {}]",
            userId, groupId, pageIndex, pageSize);
        getCategoryGroup(userId, groupId);
        Page<Category> result = categoryRepository.findByGroupId(groupId, pageIndex, pageSize);

        if (log.isDebugEnabled()) {
            log.debug("Listing categories for group [groupId: {}, page: {}, pageSize: {}, size: {}, totalCount: {}]",
                groupId, pageIndex, pageSize, result.getContentSize(), result.getTotalCount());
        }
        return result;
    }

    /**
     * Returns the identified category. Ensures the category belongs to the specified user.
     * @param userId The user ID.
     * @param categoryId The category ID.
     * @return The identified category.
     */
    public Category getCategory(UUID userId, UUID categoryId) {
        log.info("Getting category [userId: {}, categoryId: {}]", userId, categoryId);
        return validate(userId, categoryId);
    }

    public Category createCategory(UUID userId, UUID groupId, String name, String description, String colour) {
        log.info("Creating category [userId: {}, groupId: {}, name: {}]",
            userId, groupId, name);

        CategoryGroup group = getCategoryGroup(userId, groupId);

        Category category = group.addCategory(name, cat -> cat
            .description(Strings.trimOrNull(description))
            .colour(Strings.trimOrNull(colour))
        );

        categoryRepository.save(category);
        return category;
    }

    public Category updateCategory(UUID userId, UUID categoryId,
                                   String name, String description, String colour) {
        log.info("Updating category [userId: {}, categoryId: {}, name: {}]",
            userId, categoryId, name);

        String newName = Strings.trimOrNull(name);
        if (newName == null) {
            throw new MissingParameterException("Category.name");
        }

        Category category = validate(userId, categoryId);

        CategoryGroup group = category.getGroup();
        group.getCategory(newName)
            .filter(c -> !c.getId().equals(categoryId))
            .ifPresent(c -> { throw new CategoryAlreadyExistsException(c); });

        category.setName(newName);
        category.setDescription(Strings.trimOrNull(description));
        category.setColour(Strings.trimOrNull(colour));

        categoryRepository.save(category);
        return category;
    }

    /**
     * Deletes the identified category, if it belongs to the identified user.
     * All the selectors of the category will also be deleted.
     * @param userId the user attempting to delete the category.
     * @param categoryId the category's identifier.
     * @return the deleted category.
     */
    public Category deleteCategory(UUID userId, UUID categoryId) {
        log.info("Deleting category [userId: {}, categoryId: {}]",
            userId, categoryId);

        Category category = validate(userId, categoryId);
        categoryRepository.delete(category);
        return category;
    }

    public Page<CategorySelector> getCategorySelectors(UUID userId, UUID categoryId,
                                                       int pageIndex, int pageSize) {
        log.info("Listing category selectors [userId: {}, categoryId: {},, page: {}, pageSize: {}]",
            userId, categoryId, pageIndex, pageSize);

        Category category = validate(userId, categoryId);
        Page<CategorySelector> result = Page.of(category.getSelectors(), pageIndex, pageSize);

        if (log.isDebugEnabled()) {
            log.debug("Listing category selectors [userId: {}, categoryId: {}, page: {}, pageSize: {}, size: {}, totalCount: {}]",
                userId, categoryId, pageIndex, pageSize, result.getContentSize(), result.getTotalCount());
        }
        return result;

    }
    /**
     * Returns the selectors for the identified category and account.
     *
     * @param userId The user ID to whom the category group and account must belong.
     * @param categoryId The category ID.
     * @param accountId The account ID to which the selectors are associated.
     * @return The category selectors.
     */
    public Page<CategorySelector> getCategorySelectors(UUID userId, UUID categoryId, UUID accountId,
                                                       int pageIndex, int pageSize) {
        log.info("Listing category selectors [userId: {}, categoryId: {}, accountId: {}, page: {}, pageSize: {},]",
            userId, categoryId, accountId, pageIndex, pageSize);

        Category category = validate(userId, categoryId);
        validateAccount(category.getGroup(), accountId);

        Page<CategorySelector> result = Page.of(category.getSelectors().stream()
            .filter(selector -> selector.getAccountId().equals(accountId))
            .toList(), pageIndex, pageSize);

        if (log.isDebugEnabled()) {
            log.debug("Listing category selectors [userId: {}, categoryId: {}, accountId: {}, page: {}, pageSize: {}, size: {}, totalCount: {}]",
                userId, categoryId, accountId, pageIndex, pageSize, result.getContentSize(), result.getTotalCount());
        }
        return result;
    }

    /**
     * Sets the selectors for the identified category and account. Overwrites any existing selectors
     * for the same account.
     *
     * @param userId The user ID to whom the category and account must belong.
     * @param categoryId The category ID.
     * @param accountId The account ID to which the selectors are to be associated.
     * @param selectors The category selectors to be set.
     * @return the updated category selectors.
     */
    public Collection<CategorySelector> setCategorySelectors(UUID userId,
                                                             UUID categoryId,
                                                             UUID accountId,
                                                             Collection<CategorySelector> selectors) {
        log.info("Setting category selectors [userId: {}, categoryId: {}, accountId: {}]",
            userId, categoryId, accountId);

        Category category = validate(userId, categoryId);
        validateAccount(category.getGroup(), accountId);

        category.getSelectors().removeIf(selector -> selector.getAccountId().equals(accountId));
        if (selectors != null) {
            selectors.forEach(newSelector -> category.addSelector(accountId, selector -> selector
                .infoContains(Strings.trimOrNull(newSelector.getInfoContains()))
                .refContains(Strings.trimOrNull(newSelector.getRefContains()))
                .creditorContains(Strings.trimOrNull(newSelector.getCreditorContains()))
            ));
        }

        categoryRepository.save(category);
        return category.getSelectors();
    }

    /**
     * Generates the category statistics for the specified user, group and date range.
     * @param userId The user ID.
     * @param groupId The group ID.
     * @param startDate The start date (inclusive).
     * @param endDate The end date (exclusive).
     * @return The list of category statistics.
     */
    public List<CategoryStatistics> getStatistics(UUID userId, UUID groupId, Instant startDate, Instant endDate) {
        CategoryGroup group = getCategoryGroup(userId, groupId);
        return categoryGroupRepository.getStatistics(group, startDate, endDate);
    }

    /**
     * Ensures that both the category group and account exist and belong to the specified user.
     * @param userId The user ID.
     * @param accountId The account ID.
     * @return The identified category group.
     * @throws NotFoundException if the group or account does not exist or does not belong to the user.
     */
    private CategoryGroup validate(UUID userId, UUID groupId, UUID accountId) {
        CategoryGroup group = getCategoryGroup(userId, groupId);
        validateAccount(group, accountId);
        return group;
    }

    /**
     * Ensures that the category exists and belongs to the specified user.
     * @param userId The user ID.
     * @param categoryId The category ID.
     * @return The identified category.
     * @throws NotFoundException if the category does not exist or does not belong to the user.
     */
    private Category validate(UUID userId, UUID categoryId) {
        Category category = categoryRepository.findByIdOptional(categoryId)
            .orElseThrow(() -> new NotFoundException("Category", categoryId));

        CategoryGroup group = category.getGroup();
        if (!group.getUserId().equals(userId)) {
            throw new NotFoundException("Category", categoryId);
        }

        return category;
    }

    /**
     * Ensures that the identified account exists and belongs to the same user as the given group.
     * @param group The category group.
     * @param accountId The account ID.
     * @throws NotFoundException if the account does not exist or does not belong to the group's user.
     */
    private void validateAccount(CategoryGroup group, UUID accountId) {
        accountRepository.findByIdOptional(accountId)
            .filter(account -> account.getUserId().equals(group.getUserId()))
            .orElseThrow(() -> new NotFoundException("Account", accountId));
    }
}

File: client/src/services/category.service.ts
import http from './http-common';
import PaginatedList from '../model/paginated-list.model';
import { CategoryGroup, Category, CategorySelector, CategoryStatistics } from '../model/category.model';

class CategoryService {
  /**
   * Returns a page of the authenticate user's category groups. The
   * groups are ordered on their name, in ascending order.
   * @param page the zero-based index of the page to be returned.
   * @param pageSize the maximum number of category groups per page.
   * @returns the identified page of the user's category groups, in
   *  ascending name order.
   */
  getGroupsPage(page: number = 0, pageSize: number = 20): Promise<PaginatedList<CategoryGroup>> {
    return http.get<PaginatedList<CategoryGroup>>('/rails/category-groups', { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  /**
   * Returns all the configured category groups for the authenticated user.
   * The groups are ordered on their name, in ascending order.
   * @returns the user's category groups, in ascending name order. 
   */
  async fetchAllGroups(): Promise<Array<CategoryGroup>> {
    var response = await this.getGroupsPage(0, 100);
    var groups = response.items as Array<CategoryGroup>;
    while (response.links.next) {
      response = await this.getGroupsPage(response.page + 1, 100);
      groups = groups.concat(response.items);
    }
    return groups;
  }

  /**
   * Returns the identified category group, but not its categories.
   * @param groupId the category group's unique identifier.
   * @returns the identified category group.
   */
  getGroup(groupId: string): Promise<CategoryGroup> {
    return http.get<CategoryGroup>(`/rails/category-groups/${groupId}`)
      .then(response => response.data);
  }

  /**
   * Creates a new category group for the authenticated user.
   * @param group the details of the new category group.
   * @returns the newly created category group.
   */
  createGroup(group: CategoryGroup): Promise<CategoryGroup> {
    return http.post<CategoryGroup>('/rails/category-groups', group)
      .then(response => response.data);
  }

  /**
   * Modifies the mutable properties of the identified group.
   * @param group the properties of the category group to be updated,
   * including its identifier.
   * @returns the modified category group.
   */
  updateGroup(group: CategoryGroup): Promise<CategoryGroup> {
    return http.put<CategoryGroup>(`/rails/category-groups/${group.id}`, group)
      .then(response => response.data);
  }

  /**
   * Deletes the authenticated user's identified category group. All categories
   * and category selectors within the group will also be deleted.
   * @param groupId the category group's unique identifier.
   * @returns an empty response.
   */
  deleteGroup(groupId: string): Promise<any> {
    return http.delete<void>(`/rails/category-groups/${groupId}`);
  }

  /**
   * Returns a page of the categories contained within the identified
   * category group. The categories are ordered on their name, in ascending
   * order.
   * @param groupId the group identifier.
   * @param page the zero-based index of the page to be returned.
   * @param pageSize the maximum number of category groups per page.
   * @returns the identified page of the identified group's categories.
   */
  getCategoriesPage(groupId: string, page: number = 0, pageSize: number = 20): Promise<PaginatedList<Category>> {
    return http.get<PaginatedList<Category>>(`/rails/category-groups/${groupId}/categories`, { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  /**
   * Returns all the categories contained within the identified category
   * group. The categories are ordered on their name, in ascending order.
   * @param groupId the group identifier.
   * @returns the identified group's categories, in ascending name order.
   */
  async fetchAllCategories(groupId: string): Promise<Array<Category>> {
    var response = await this.getCategoriesPage(groupId, 0, 100);
    var categories = response.items as Array<Category>;
    while (response.links.next) {
      response = await this.getCategoriesPage(groupId, response.page + 1, 100);
      categories = categories.concat(response.items);
    }
    return categories;
  }

  /**
   * Returns the identified category.
   * @param categoryId the category's unique identifier.
   * @returns the identified category.
   */
  getCategory(categoryId: string): Promise<Category> {
    return http.get<Category>(`/rails/categories/${categoryId}`)
      .then(response => response.data);
  }

  /**
   * Creates a new category within the identified category group, for the
   * authenticated user.
   * @param groupId the unique identifier of the category group.
   * @param category the details of the new category.
   * @returns the newly created category.
   */
  createCategory(groupId: string, category: Category): Promise<Category> {
    return http.post<Category>(`/rails/category-groups/${groupId}/categories`, category)
      .then(response => response.data);
  }

  /**
   * Modifies the mutable properties of the identified category.
   * @param category the properties of the category to be updated, including
   *  its identifier.
   * @returns the modified category.
   */
  updateCategory(category: Category): Promise<Category> {
    return http.put<Category>(`/rails/categories/${category.id}`, category)
      .then(response => response.data);
  }

  /**
   * Deletes the authenticated user's identified category. All the selectors of
   * the category will also be deleted.
   * @param categoryId the category's unique identifier.
   * @returns an empty response.
   */
  deleteCategory(categoryId: string): Promise<any> {
    return http.delete<void>(`/rails/categories/${categoryId}`);
  }

  /**
   * Returns the category selectors for the identified category, across all accounts.
   * @param categoryId the category's unique identifier.
   * @param page the zero-based index of the page to be returned.
   * @param pageSize the maximum number of selectors per page.
   * @returns those category selectors that belong to the category.
   */
  getCategorySelectors(categoryId: string,page: number = 0, pageSize: number = 20): Promise<PaginatedList<CategorySelector>> {
    return http.get<PaginatedList<CategorySelector>>(`/rails/categories/${categoryId}/selectors`, { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  async getAllCategorySelectors(categoryId: string): Promise<Array<CategorySelector>> {
    var response = await this.getCategorySelectors(categoryId, 0, 100);
    var selectors = response.items as Array<CategorySelector>;
    while (response.links.next) {
      response = await this.getCategorySelectors(categoryId, response.page + 1, 100);
      selectors = selectors.concat(response.items);
    }
    return selectors;
  }

  /**
   * Moves the identified category selector to the identified destination category.
   * @param selector the category selector to be moved.
   * @param destCategoryId the identifier of the category to which the selector is to be moved.
   * @returns 
   */
  moveCategorySelector(selector: CategorySelector, destCategoryId: string): Promise<CategorySelector> {
    const destination = {
      categoryId: destCategoryId
    }
    return http.put<CategorySelector>(`/rails/categories/${selector.categoryId}/selectors/${selector.id!}`, destination)
      .then(response => response.data);
  }

  /**
   * Deletes the identified category selector.
   * @param selector the selector to be deleted.
   * @returns an empty response.
   */
  deleteCategorySelector(selector: CategorySelector): Promise<any> {
    return http.delete<void>(`/rails/categories/${selector.categoryId}/selectors/${selector.id!}`);
  }

  /**
   * Returns the category selectors for both the identified category and account.
   * @param categoryId the category's unique identifier.
   * @param accountId the account's unique identifier.
   * @param page the zero-based index of the page to be returned.
   * @param pageSize the maximum number of selectors per page.
   * @returns those category selectors that belong to both the category and account.
   */
  getAccountCategorySelectors(categoryId: string, accountId: string,
                              page: number = 0, pageSize: number = 20): Promise<PaginatedList<CategorySelector>> {
    return http.get<PaginatedList<CategorySelector>>(`/rails/categories/${categoryId}/account-selectors/${accountId}`, { params: { "page": page, "page-size": pageSize }})
      .then(response => response.data);
  }

  async getAllAccountCategorySelectors(categoryId: string, accountId: string): Promise<Array<CategorySelector>> {
    var response = await this.getAccountCategorySelectors(categoryId, accountId, 0, 100);
    var selectors = response.items as Array<CategorySelector>;
    while (response.links.next) {
      response = await this.getAccountCategorySelectors(categoryId, accountId, response.page + 1, 100);
      selectors = selectors.concat(response.items);
    }
    return selectors;
  }

  /**
   * Replaces the category selectors of the identified category and account with
   * those provided in the given array.
   * @param categoryId the category's unique identifier.
   * @param accountId the account's unique identifier.
   * @param selectors the collection of category selectors to be assigned.
   * @returns 
   */
  setAccountCategorySelectors(categoryId: string, accountId: string,
                              selectors: Array<CategorySelector>): Promise<Array<CategorySelector>> {
    return http.put<Array<CategorySelector>>(`/rails/categories/${categoryId}/account-selectors/${accountId}`, selectors)
      .then(response => response.data);
  }

  /**
   * Returns the statistics for the transactions covered by the given category
   * group over the given date range. The response will contain an entry for
   * each of the group's categories for which transactions exist within the
   * given date range, across all accounts.
   * @param groupId the category group's unique identifier.
   * @param fromDate the date, inclusive, from which the transaction should start.
   * @param toDate the date, exclusive, to which the transaction should end.
   * @returns an array of the category group's transaction statistics.
   */
  getCategoryStatistics(groupId: string, fromDate: Date, toDate: Date): Promise<Array<CategoryStatistics>> {
    const fromDateStr = fromDate.toISOString().substring(0, 10);
    const toDateStr = toDate.toISOString().substring(0, 10);
    return http.get<Array<CategoryStatistics>>(`/rails/category-groups/${groupId}/statistics`,
      { params: { "from-date": fromDateStr, "to-date": toDateStr }})
      .then(response => response.data);
  }
}

const instance = new CategoryService();
export default instance;
